### ISSUE TITLE
Rebuild angular universal on save

### ISSUE THREAD

**Issue by samialmorshedi** [angular/universal#1022]
I turned my application to angular universal. It's great but I have a problem with developing mode. I need see my editing result real-time. I need restart my app on each atrl+s (save) like `"build --watch"` and `"nodmeon"`.

My server is expressjs and I'm using `"npm run build:ssr"` after each edit. It takes lots of time!!

How can I do that?

**Comment by hiepxanh** [comment]
agree

**Comment by Toxicable** [comment]
We have some advise for this comming, will update when we have more

**Comment by avik-so** [comment]
Any updates on this?

**Comment by sushrest** [comment]
Agree with @samialmorshedi, I am managing this with nodemon until the formal advice. 

Add Following line on package.json
`"ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec 'npm run build:ssr && npm run serve:ssr'"`

Start App 
`npm run ssr:serve`

**Comment by alon-tsabari** [comment]
Thanks a lot!

**Comment by hiepxanh** [comment]
nice, love it will be intergrate in angular cli soon

**Comment by th0r** [comment]
@Toxicable Are there any news guys? I think [this solution](https://github.com/angular/universal/issues/1022#issuecomment-462092440) using `nodemon` is not suitable for large projects as it completely rebuilds the whole app on any change.
Are there any other approaches?

**Comment by nivethaKumar** [comment]
> pm run ssr:serve



> Agree with @samialmorshedi, I am managing this with nodemon until the formal advice.
> 
> Add Following line on package.json
> `"ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec 'npm run build:ssr && npm run serve:ssr'"`
> 
> Start App
> `npm run ssr:serve`

it's not working for me  ..getting this error "''npm' is not recognized as an internal or external command,
operable program or batch file.
[nodemon] app crashed - waiting for file changes before starting...
"

**Comment by JustasKuizinas** [comment]
> Agree with @samialmorshedi, I am managing this with nodemon until the formal advice.
> 
> Add Following line on package.json
> `"ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec 'npm run build:ssr && npm run serve:ssr'"`
> 
> Start App
> `npm run ssr:serve`

I get 
```
''npm' is not recognized as an internal or external command,
operable program or batch file.
```
why?

**Comment by almgwary** [comment]
Any updates we need this feature urgently, hot reload on development with NodeJs ssr

**Comment by hiepxanh** [comment]
they asking for what feature should be improve but they didn't even check the current problem

**Comment by almgwary** [comment]
@hiepxanh  we need command like `ng serve` for  SSR to run the app under nodejs with live reload.

**Comment by almgwary** [comment]
@MarkPieszak  any suggestions to run SSR on development with live reload like `ng serve`

**Comment by Adam23** [comment]
Any updates we need this feature urgently, hot reload on development with ssr

**Comment by static2000** [comment]
> Any updates we need this feature urgently, hot reload on development with NodeJs ssr

Agree with @almgwary ,Especially for large projects,  it takes lots of time now.

**Comment by davidecampello** [comment]
i am interested too

**Comment by KingDarBoja** [comment]
Just letting my comment to watch out for any release of this tool, It is a bit annoying to rebuild everytime I perform a single change on my app. 

Cheers!

**Comment by MirzaLeka** [comment]
> > Agree with @samialmorshedi, I am managing this with nodemon until the formal advice.
> > Add Following line on package.json
> > `"ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec 'npm run build:ssr && npm run serve:ssr'"`
> > Start App
> > `npm run ssr:serve`
> 
> I get
> 
> ```
> ''npm' is not recognized as an internal or external command,
> operable program or batch file.
> ```
> 
> why?

I had the same issue with single quote and fixed it using the double. Just make sure you use escape characters. Try this instead
`"ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec \"npm run build:ssr && npm run serve:ssr\" "`

Make sure you have nodemon installed in the project or at least have it installed global.
I'm using this script and before you ask, yes it works, but compiler is still slow.
`    "ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec \"npm run buildAndServe:ssr\""`

**Comment by RedactedProfile** [comment]
Not to sound too pushy, but this issue was submitted on June 11th 2018, and here we are November 19th 2019 with no clear path about what to do other than "just run `build:ssr` and then `serve:ssr` over and over again, regardless of project size". 

My problem right now is that even `ng serve` is broken when going with a fresh ngUniversal Express engine schematic (only produced a white page, with an error in the compiled vendor js about the `domino` dependency).  So if im building from scratch to what is going to be a pretty sizable project, im potentially wasting ridiculous amounts of dev time to watching universal compile out before being able to test my pages

**Comment by alan-agius4** [comment]
Hi all, in the last RC version we have added this feature in `@nguniversal/builders` package which.

To try up out you need to update to the latest RC which will add the builder and related configuration to your provide via a migration.

**Express**
```bash
ng update @angular/cli @angular/core @nguniversal/express-engine --next
```

**Hapi**
```bash
ng update @angular/cli @angular/core @nguniversal/hapi-engine --next
```

The builder can be executed via:
```
ng run <projectName>:serve-ssr
```

or 
```
yarn serve:ssr:dev
npm run serve:ssr:dev
```

**Comment by laurencefass** [comment]
seeing message **Repository is not clean. Please commit or stash any changes before updating.**

```
bash-5.0# ng update @angular/cli @angular/core @nguniversal/express-engine --next
The installed Angular CLI version is older than the latest pre-release version.
Installing a temporary version to perform the update.
Installing packages for tooling via npm.
Installed packages for tooling via npm.
Repository is not clean. Please commit or stash any changes before updating.

```
I had to --allow-dirty to get it to build.

**Comment by laurencefass** [comment]
Can anyone advise how to change the port on the dev server port from 4200.

**Comment by laurencefass** [comment]
ive tested this and its not usable. appears to do a major recompilation for each small change and still not updating the browser.

also generates EADDRESSINUSE  error

after saving source code changes there is a 30 second compilation step and i have to manually reload the browser to see the change.

```
Hash: 6e30d403d281f596bd9c
Time: 18935ms
Built at: 12/17/2019 1:04:07 PM
      Asset      Size  Chunks                          Chunk Names
    main.js  6.38 MiB    main  [emitted]        [big]  main
main.js.map  6.91 MiB    main  [emitted] [dev]         main
Entrypoint main [big] = main.js main.js.map
chunk {main} main.js, main.js.map (main) 6.02 MiB [entry] [rendered]

Compiled successfully.

Date: 2019-12-17T13:04:50.203Z - Hash: 2c71f08bdf2acdc697dc
4 unchanged chunks
chunk {main} main.js, main.js.map (main) 72.8 kB [initial] [rendered]
Time: 9084ms
Hash: 7a89cefb5aa17da58743
Time: 7750ms
Built at: 12/17/2019 1:04:50 PM
      Asset      Size  Chunks                          Chunk Names
    main.js  6.38 MiB    main  [emitted]        [big]  main
main.js.map  6.91 MiB    main  [emitted] [dev]         main
Entrypoint main [big] = main.js main.js.map
chunk {main} main.js, main.js.map (main) 6.02 MiB [entry] [rendered]
95% emitting
Compiled successfully.
/app/angular/ssr/dist/server/main.js:166079
                            throw error;
                            ^

Error: listen EADDRINUSE: address already in use :::45287
    at Server.setupListenHandle [as _listen2] (net.js:1308:16)
    at listenInCluster (net.js:1356:12)
    at Server.listen (net.js:1444:7)
    at Function.listen (/app/angular/ssr/dist/server/main.js:136360:24)
    at run (/app/angular/ssr/dist/server/main.js:168392:12)
    at Object../server.ts (/app/angular/ssr/dist/server/main.js:168398:5)
    at __webpack_require__ (/app/angular/ssr/dist/server/main.js:20:30)
    at Object.0 (/app/angular/ssr/dist/server/main.js:168951:18)
    at __webpack_require__ (/app/angular/ssr/dist/server/main.js:20:30)
    at /app/angular/ssr/dist/server/main.js:84:18
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1335:8)
    at ZoneDelegate.invokeTask (/app/angular/ssr/dist/server/main.js:166308:35)
    at Zone.runTask (/app/angular/ssr/dist/server/main.js:166075:51)
    at ZoneTask.invokeTask (/app/angular/ssr/dist/server/main.js:166390:38)
    at ZoneTask.invoke (/app/angular/ssr/dist/server/main.js:166379:52)
    at data.args.<computed> (/app/angular/ssr/dist/server/main.js:167330:63)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  code: 'EADDRINUSE',
  errno: -98,
  syscall: 'listen',
  address: '::',
  port: 45287
}


Date: 2019-12-17T13:05:06.761Z - Hash: 1b5177c4533831095b50
4 unchanged chunks
chunk {main} main.js, main.js.map (main) 72.8 kB [initial] [rendered]
Time: 13288ms
Hash: 2a062acd0fef5181acfd
Time: 11399ms
Built at: 12/17/2019 1:05:06 PM
      Asset      Size  Chunks                          Chunk Names
    main.js  6.38 MiB    main  [emitted]        [big]  main
main.js.map  6.91 MiB    main  [emitted] [dev]         main
Entrypoint main [big] = main.js main.js.map
chunk {main} main.js, main.js.map (main) 6.02 MiB [entry] [rendered]

Compiled successfully.
```

**Comment by laurencefass** [comment]
Is the performance of this recompilation an initial offering? Can and will it be improved? I was very surprised to see such a lengthy compilation step. Razzle.js and Next.js deliver page changes almost instantly. Why is Angular SSR so slow?

**Comment by laurencefass** [comment]
Another question if i may:

Previously (A8) i was running my dev server behind a reverse proxy with     

"start": "ng serve --host 0.0.0.0 --port 3604 --disable-host-check --public-host https://mydomain/ssr",
 
now im running with 

"dev:ssr": "ng run default:serve-ssr --host 0.0.0.0 --port 3604"

and including --public-host in my dev:ssr command results in error.

```
Unknown option: '--public-host'
Unknown option: 'https://angular.syntapse.co.uk/ssr'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! default@0.0.0 dev:ssr: `ng run default:serve-ssr --host 0.0.0.0 --port 3604 --public-host https://angular.syntapse.co.uk/ssr`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the default@0.0.0 dev:ssr script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-12-17T13_24_58_767Z-debug.log
```

how do i replace replace --public-host in Angular 9?

**Comment by alan-agius4** [comment]
Hi @laurencefass, thanks for trying this out.

Just a couple of notes;
- This by no means is a replacement for `ng serve`, `ng serve` is faster as it only has a single compilation. The client bundle, while this SSR dev server builder has to generate both server and client bundle. Also there other features that the SSR dev sever builder doesn't support. One of them being HMR.
- `public-host` is currently not available, kindly file a feature request.
- On every change both client and server needs to be compilations. We'll be looking in way to optimize it in the near future.
- With regards to the `listen EADDRINUSE: address already in use`, error can you open a new issue and provide the output of `ng version`? Thanks.

**Comment by laurencefass** [comment]
thanks alan. as a workaround is there another way to set the --public-host to allow paths beyond default /

**Comment by maddy2894** [comment]
> Agree with @samialmorshedi, I am managing this with nodemon until the formal advice.
> 
> Add Following line on package.json
> `"ssr:serve": "nodemon --watch src/ -e ts,html,scss --exec 'npm run build:ssr && npm run serve:ssr'"`
> 
> Start App
> `npm run ssr:serve`

This works only if there are any changes to the angular app, I have client and server both under same directory, is there any way to compile (hot reload) on changes made to nodeJs/express server backend files?

**Comment by KingDarBoja** [comment]
> thanks alan. as a workaround is there another way to set the --public-host to allow paths beyond default /

There is some work in progress to provide such option on #1401 thanks to your suggestion 🎉

**Comment by CaerusKaru** [comment]
@alan-agius4 @manfredsteyer Can this be closed?

**Comment by alan-agius4** [comment]
@CaerusKaru, yes thanks.
